File: ts/src/accountKey.ts
import { ml_kem768 } from "@noble/post-quantum/ml-kem";
import { arrayBufferToBase64, base64ToArrayBuffer, keyHash } from "./utils.ts";
import { isValidMasterKeyPrivate, isValidMasterKeyPublic, signMasterKey } from "./masterKey.ts";
import type { accountKey, EncryptedData } from "./type.ts";

export async function generateAccountKey(masterKey: {
  publicKey: string;
  privateKey: string;
}): Promise<
  {
    publickKey: string;
    privateKey: string;
    sign: string;
  } | null
> {
  if (!isValidMasterKeyPrivate(masterKey.privateKey)) {
    return null;
  }
  if (!isValidMasterKeyPublic(masterKey.publicKey)) {
    return null;
  }
  const key = ml_kem768.keygen();
  const publicKeyBinary = arrayBufferToBase64(
    key.publicKey as unknown as ArrayBuffer,
  );
  const privateKeyBinary = arrayBufferToBase64(
    key.secretKey as unknown as ArrayBuffer,
  );
  const timestamp = new Date().getTime();
  const publickKeyObj: accountKey = {
    keyType: "accountKeyPublic",
    key: publicKeyBinary,
    algorithm: "ML-KEM-768",
    timestamp: timestamp,
  };
  const privateKeyObj: accountKey = {
    keyType: "accountKeyPrivate",
    key: privateKeyBinary,
    algorithm: "ML-KEM-768",
    timestamp: timestamp,
  };
  const publickKey = JSON.stringify(publickKeyObj);
  const privateKey = JSON.stringify(privateKeyObj);
  const sign = signMasterKey(
    masterKey.privateKey,
    publickKey,
    await keyHash(masterKey.publicKey),
  );
  if (!sign) {
    return null;
  }
  return {
    publickKey: publickKey,
    privateKey: privateKey,
    sign: sign,
  };
}

export function isValidAccountKeyPublic(key: string): boolean {
  if (key.length !== 1645) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "accountKeyPublic") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 1184) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export function isValidAccountKeyPrivate(key: string): boolean {
  if (key.length !== 3266) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "accountKeyPrivate") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 2400) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export async function encryptDataAccountKey(
  key: string,
  data: string,
): Promise<string | null> {
  if (!isValidAccountKeyPublic(key)) {
    return null;
  }
  const { key: keyBinary } = JSON.parse(key);
  const dataArray = new TextEncoder().encode(data);
  const ciphertext = ml_kem768.encapsulate(
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
    new Uint8Array(32),
  );
  const ciphertextString = arrayBufferToBase64(
    ciphertext.cipherText as unknown as ArrayBuffer,
  );
  const keyHashString = await keyHash(key);
  const importedKey = await crypto.subtle.importKey(
    "raw",
    new Uint8Array(ciphertext.sharedSecret),
    "AES-GCM",
    true,
    ["encrypt", "decrypt"],
  );
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const encryptedData = await crypto.subtle.encrypt(
    {
      name: "AES-GCM",
      iv: iv,
    },
    importedKey,
    dataArray,
  );
  const viString = arrayBufferToBase64(iv as unknown as ArrayBuffer);
  const encryptedDataString = arrayBufferToBase64(encryptedData);
  const result: EncryptedData = {
    encryptedData: encryptedDataString,
    iv: viString,
    cipherText: ciphertextString,
    keyType: "accountKey",
    keyHash: keyHashString,
    algorithm: "AES-GCM",
  };
  return JSON.stringify(result);
}

export function isValidEncryptedDataAccountKey(data: string): boolean {
  const { keyType, keyHash, iv, cipherText, algorithm } = JSON.parse(data);
  const sha256 = new Uint8Array(base64ToArrayBuffer(keyHash));
  if (keyType !== "accountKey") {
    return false;
  }
  if (sha256.length !== 32) {
    return false;
  }
  if (new Uint8Array(base64ToArrayBuffer(iv)).length !== 12) {
    return false;
  }
  if (new Uint8Array(base64ToArrayBuffer(cipherText)).length !== 1088) {
    return false;
  }
  return true;
}

export async function decryptDataAccountKey(
  key: string,
  data: string,
): Promise<string | null> {
  if (!isValidAccountKeyPrivate(key)) {
    return null;
  }
  if (!isValidEncryptedDataAccountKey(data)) {
    return null;
  }
  const { key: keyBinary } = JSON.parse(key);
  const { encryptedData: binaryEncryptedData, iv, cipherText } = JSON.parse(
    data,
  );
  const sharedSecret = ml_kem768.decapsulate(
    new Uint8Array(base64ToArrayBuffer(cipherText)),
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
  );
  const importedKey = await crypto.subtle.importKey(
    "raw",
    new Uint8Array(sharedSecret),
    "AES-GCM",
    true,
    ["encrypt", "decrypt"],
  );
  const decryptedData = await crypto.subtle.decrypt(
    {
      name: "AES-GCM",
      iv: new Uint8Array(base64ToArrayBuffer(iv)),
    },
    importedKey,
    new Uint8Array(base64ToArrayBuffer(binaryEncryptedData)),
  );
  return new TextDecoder().decode(decryptedData);
}

export function isValidEncryptedAccountKey(data: string): boolean {
  if (!isValidEncryptedDataShareKey(data)) {
    return false;
  }
  if (data.length !== 5966) {
    console.log(data.length);
    return false;
  }
  return true;
}

export function isValidEncryptedDataShareKey(data: string): boolean {
  const { keyType, keyHash, iv, cipherText, algorithm } = JSON.parse(data);
  const sha256 = new Uint8Array(base64ToArrayBuffer(keyHash));
  if (keyType !== "shareKey") {
    return false;
  }
  if (sha256.length !== 32) {
    return false;
  }
  if (new Uint8Array(base64ToArrayBuffer(iv)).length !== 12) {
    return false;
  }
  if (new Uint8Array(base64ToArrayBuffer(cipherText)).length !== 1088) {
    return false;
  }
  return true;
}

File: rust/src/account_key.rs
use std::time::{SystemTime, UNIX_EPOCH};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use serde_json;
use crate::r#type::{AccountKey, EncryptedData};
use crate::utils::key_hash;
use crate::crypto::{encrypt, decrypt};
use crate::master_key::sign_master_key;
use crate::keyutils::is_valid_kem_key;

/// アカウント鍵ペア生成 (JSON文字列＋署名)
pub fn generate_account_key(
    master_public_json: &str,
    master_private_json: &str,
) -> Option<(String, String, String)> {
    if !crate::master_key::is_valid_master_key_public(master_public_json)
        || !crate::master_key::is_valid_master_key_private(master_private_json)
    {
        return None;
    }
    let (pub_b64, priv_b64) = crate::keyutils::generate_kem_key_pair().ok()?;
    #[cfg(target_arch = "wasm32")]
    let timestamp = 0u64;
    #[cfg(not(target_arch = "wasm32"))]
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH).ok()?
        .as_millis() as u64;
    let pub_obj = AccountKey {
        key_type: "accountKeyPublic".into(),
        key: pub_b64.clone(),
        algorithm: "ML-KEM-768".into(),
        timestamp,
    };
    let priv_obj = AccountKey {
        key_type: "accountKeyPrivate".into(),
        key: priv_b64.clone(),
        algorithm: "ML-KEM-768".into(),
        timestamp,
    };
    let pub_json = serde_json::to_string(&pub_obj).ok()?;
    let priv_json = serde_json::to_string(&priv_obj).ok()?;
    let mh = key_hash(master_public_json);
    let sign = sign_master_key(master_private_json, &pub_json, &mh)?;
    Some((pub_json, priv_json, sign))
}

/// 公開鍵 JSON 検証
pub fn is_valid_account_key_public(json: &str) -> bool {
    if let Ok(ak) = serde_json::from_str::<AccountKey>(json) {
        ak.key_type == "accountKeyPublic"
            && ak.algorithm == "ML-KEM-768"
            && is_valid_kem_key(&ak.key, true)
    } else {
        false
    }
}

/// 秘密鍵 JSON 検証
pub fn is_valid_account_key_private(json: &str) -> bool {
    if let Ok(ak) = serde_json::from_str::<AccountKey>(json) {
        ak.key_type == "accountKeyPrivate"
            && ak.algorithm == "ML-KEM-768"
            && is_valid_kem_key(&ak.key, false)
    } else {
        false
    }
}

/// アカウント鍵による暗号化 (EncryptedData JSON)
pub fn encrypt_data_account_key(
    key_json: &str,
    data: &str,
) -> Option<String> {
    if !is_valid_account_key_public(key_json) {
        return None;
    }
    let ak: AccountKey = serde_json::from_str(key_json).ok()?;
    let enc = encrypt(data, &ak.key);
    let ed = EncryptedData {
        key_type: "accountKey".into(),
        key_hash: key_hash(key_json),
        encrypted_data: enc.encrypted_data,
        iv: enc.iv,
        algorithm: Some(enc.algorithm),
        cipher_text: Some(enc.cipher_text),
    };
    serde_json::to_string(&ed).ok()
}

/// EncryptedData JSON 検証
pub fn is_valid_encrypted_data_account_key(json: &str) -> bool {
    if let Ok(ed) = serde_json::from_str::<EncryptedData>(json) {
        ed.key_type == "accountKey"
            && ed.algorithm.as_deref() == Some("AES-GCM")
            && BASE64.decode(&ed.key_hash).map(|v| v.len() == 32).unwrap_or(false)
            && BASE64.decode(&ed.iv).map(|v| v.len() == 12).unwrap_or(false)
            && BASE64.decode(&ed.encrypted_data).is_ok()
            && ed.cipher_text.as_ref()
                .and_then(|ct| BASE64.decode(ct).ok())
                .is_some()
    } else {
        false
    }
}

/// アカウント鍵による復号
pub fn decrypt_data_account_key(
    key_json: &str,
    encrypted_json: &str,
) -> Option<String> {
    if !is_valid_account_key_private(key_json)
        || !is_valid_encrypted_data_account_key(encrypted_json)
    {
        return None;
    }
    let ak: AccountKey = serde_json::from_str(key_json).ok()?;
    let ed: EncryptedData = serde_json::from_str(encrypted_json).ok()?;
    let ciphertext = ed.cipher_text.as_ref()?;
    Some(decrypt(&ed.encrypted_data, ciphertext, &ed.iv, &ak.key))
}

/// EncryptedAccountKey 検証 (エイリアス)
pub fn is_valid_encrypted_account_key(json: &str) -> bool {
    is_valid_encrypted_data_account_key(json)
}
